src/asm/trampoline-arm.S

```diff
@@ -24,12 +24,15 @@
 #include "config.h"
 #include "mangling.h"
 
+.arm
+
 #ifdef ASSEMBLER_SUPPORTS_HIDDEN
 .hidden MANGLE(bxfi_trampoline)
 #endif
 .globl MANGLE(bxfi_trampoline)
 MANGLE(bxfi_trampoline):
-    ldr     pc, addr_data
+    ldr     r3, addr_data
+    bx      r3
```

Comment:
I wonder if bx is actually necessary, both the thumb and non-thumb trampolines should just jump without doing anything else. I suspect you switched to bx to support the thumb mode switch, but we should already be using the thumb trampoline if the target function is thumb and vice-versa, so we should be able to drop bx?

Reply:
Rereading this, I think I'm wrong: the whole point is that the function pointer in addr_data might be thumb while the original main is non-thumb. Therefore, bx is needed to seamlessly switch modes.

Reply:
Yeah, my intention was to support a ARM main with a Thumb new_main, or vice versa.